File: CrazyBot.java
import java.util.*;

public class CrazyBot {
  public static int N;
  public static int EAST;
  public static int WEST;
  public static int SOUTH;
  public static int NORTH;
  public static int[][] dist;

  public double getProbability(int n, int east, int west, int south, int north) {
    N = n;
    EAST = east;
    WEST = west;
    SOUTH = south;
    NORTH = north;
    dist = new int[][] { { 1, 0, EAST }, { -1, 0, WEST }, { 0, 1, SOUTH }, { 0, -1, NORTH } };

    Set<String> set = new HashSet<>();
    set.add("0,0");
    return dfs(1, 0, 0, 0, set);
  }

  public double dfs(double prob, int n, int x, int y, Set<String> set) {
    if (n == N)
      return prob;

    double ret = .0;

    for (int i = 0; i < 4; i++) {
      int[] d = dist[i];
      int dx = d[0];
      int dy = d[1];
      int p = d[2];

      int nx = x + dx;
      int ny = y + dy;
      String v = Integer.toString(nx) + "," + Integer.toString(ny);

      if (!set.contains(v)) {
        set.add(v);
        ret += dfs(prob * p / 100, n + 1, nx, ny, set);
        set.remove(v);
      }
    }

    return ret;
  }

  public static void main(String[] main) {
    System.out.println(new CrazyBot().getProbability(14, 25, 25, 25, 25));
  }
}

File: crazy_bot.cpp
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

class CrazyBot
{
public:
  bool visited[100][100] = {{false}};
  double prob[4];
  int dx[4] = {0, 0, -1, 1};
  int dy[4] = {-1, 1, 0, 0};
  double getProbability(int n, int east, int west, int south, int north)
  {
    prob[0] = north / 100.0;
    prob[1] = south / 100.0;
    prob[2] = west / 100.0;
    prob[3] = east / 100.0;

    return dfs(50, 50, n);
  }

  double dfs(int x, int y, int n)
  {
    if (visited[y][x])
    {
      return 0;
    }
    if (n == 0)
    {
      return 1;
    }

    visited[y][x] = true;

    double ret = 0;
    for (int i = 0; i < 4; i++)
    {
      ret += dfs(x + dx[i], y + dy[i], n - 1) * prob[i];
    }

    visited[y][x] = false;

    return ret;
  }
};

int main()
{
  CrazyBot *c = new CrazyBot;
  cout << c->getProbability(14, 25, 25, 25, 25) << endl;
}